mdbook-tera-backend/src/tera_renderer/renderer.rs

```diff
@@ -0,0 +1,152 @@
+use super::CustomComponent;
+use anyhow::{anyhow, Result};
+use lol_html::html_content::ContentType;
+use lol_html::{element, RewriteStrSettings};
+use mdbook::renderer::RenderContext;
+use serde_json::to_value;
+use std::collections::{BTreeMap, HashMap};
+use std::fs;
+use std::io::{Read, Write};
+use std::path::Path;
+use std::sync::Arc;
+use tera::Tera;
+
+pub(crate) struct Renderer {
+    ctx: Arc<RenderContext>,
+    serialized_ctx: serde_json::Value,
+    components: Vec<CustomComponent>,
+    counter: u64,
+    tera_template: Tera,
+}
+
+impl Renderer {
+    pub(crate) fn new(ctx: RenderContext, tera_template: Tera) -> Result<Renderer> {
+        let mut renderer = Renderer {
+            serialized_ctx: serde_json::to_value(&ctx)?,
+            ctx: Arc::new(ctx),
+            components: Vec::new(),
+            counter: 0,
+            tera_template,
+        };
+        renderer
+            .tera_template
+            .register_function("get_context", renderer.create_get_context_function());
+        Ok(renderer)
+    }
+
+    pub(crate) fn add_component(&mut self, mut component: CustomComponent) {
+        component.register_function("get_context", self.create_get_context_function());
+        self.components.push(component);
+    }
+
+    fn create_get_context_function(&self) -> impl tera::Function {
+        let ctx_rc = Arc::clone(&self.ctx);
+        move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+            let key = args
+                .get("key")
+                .ok_or_else(|| tera::Error::from(format!("No key argument provided")))?
+                .as_str()
+                .ok_or_else(|| {
+                    tera::Error::from(format!("Key has invalid type, expected string"))
+                })?;
+            let value = ctx_rc
+                .config
+                .get(key)
+                .ok_or_else(|| tera::Error::from(format!("Could not find key {key} in config")))?;
+            let value = to_value(value)?;
+            Ok(value)
+        }
+    }
+
+    pub(crate) fn render_book(&mut self) -> Result<()> {
+        let dest_dir = &self
+            .ctx
+            .destination
+            .parent()
+            .ok_or_else(|| {
+                anyhow!(
+                    "Destination directory {:?} has no parent",
+                    self.ctx.destination
+                )
+            })?
+            .to_owned();
```

Comment:
Can we just `unwrap()` and also avoid the `to_owned()`? It looks like a lot of code for a simple operation.

Reply:
to_owned was needed because it borrows from self otherwise